src/lib/executeSourceTransfer.js

```diff
@@ -23,19 +30,55 @@ function * executeSourceTransfer (destinationTransfer, fulfillment, ledgers, bac
   // TODO check the timestamp on the response from the ledger against
   // the transfer's expiry date
   // See https://github.com/interledgerjs/five-bells-ledger/issues/149
-  yield ledgers.getPlugin(sourceTransferLedger)
-    .fulfillCondition(sourceTransferID, fulfillment)
-    .then(() =>
+  const plugin = ledgers.getPlugin(sourceTransferLedger)
+
+  const sourceTransferExpiry = moment(destinationTransfer.expiresAt, moment.ISO_8601)
+    .add(config.expiry.minMessageWindow, 'seconds')
+  const retryTimeout = sourceTransferExpiry.diff(moment())
+
+  const timeoutPromise = new Promise((resolve, reject) => {
+    setTimeout(function () {
+      reject(`Fulfillment of source transfer ${sourceTransferID} timed out.`)
+    }, retryTimeout)
+  })
+
+  const fulfillPromise = promiseRetry(retryOpts, function (retry, number) {
+    return Promise.resolve(plugin.fulfillCondition(sourceTransferID, fulfillment))
```

Comment:
It does not always return a promise, e.g. it may return `undefnied`, hence I use `Promise.resolve` here.